### ISSUE TITLE
[Feature] Support callstack capturing for cpu scheduling events

### ISSUE THREAD

**Issue by ivberg** [google/perfetto#177]
It would be really nice for Perfetto to support optional call stack capturing for cpu scheduling events (sched_switch, sched_wakeup, sched_waking). This would be used to get more detail about what code is waiting or alerting threads. This should allow deeper root causing of issues, rather than just relying on process/thread & other flow / slice events being logged for the threads in question.

It appears Perfetto already has callstack support in the data format, at least for CPU profiling / callstack sampling. So likely the callstacks could be saved in the same table.
https://perfetto.dev/docs/reference/trace-config-proto#PerfEventConfig.CallstackSampling
https://perfetto.dev/docs/analysis/sql-tables#callstack-profilers

I don't see any reference to this callstack on cpu scheduling events currently supported for Linux FTrace:
https://perfetto.dev/docs/reference/trace-config-proto#FtraceConfig

Nor this feature request is not mentioned here in this link; although it does outline there is already support for showing the process & thread for the wakeup which is nice:
https://perfetto.dev/docs/data-sources/cpu-scheduling

For reference:
It appears this capability exists in Linux as it already exists in LTTng both user-mode and kerne-mode, so in theory Perfetto could support this as well.
See https://lttng.org/docs/v2.13/#doc-adding-context
https://lttng.org/man/1/lttng-add-context/v2.13/
https://archive.eclipse.org/tracecompass/doc/stable/org.eclipse.tracecompass.doc.user/LTTng-Kernel-Analysis.html (Process Wait Analysis)

Callstack capture on scheduling events is widely/commonly used in detailed Windows perf analysis workflows. Windows supports callstack being configured on individual kernel events; and it would be nice to have similar functionality in a generic way in order to pick and choose stackwalk individual events.
https://docs.microsoft.com/en-us/windows-hardware/test/wpt/stackwalk

**Comment by primiano** [comment]
It actually does if you combine the callstack sampler (Traced_perf) with tracepoints.

It's an extremely new feature (the all "linux.perf" data source is undocumented).
I will leave @rsavitski to comment more, but I think you can do what you want with something like this on Linux/Android:

```
      data_sources: {
        config {
          name: "linux.perf"
          perf_event_config {
            timebase {
              tracepoint {
                name: "sched:sched_switch"
                filter: "dirty == 0"
              }
              frequency: 50
            }
            callstack_sampling {
              kernel_frames: true
            }
          }
        }
      }
```

**Comment by ivberg** [comment]
Very nice @primiano and thx for the quick reply! We will have to try this out; and of course it would be great if this were officially documented further. Can even make a stretch wish to be supported in the Perfetto UI (Record New Trace).

Questions: 
1. When you say extremely new feature, w.r.t trace capture does that imply a certain version of Android or Linux?
2. How does the frequency fit into this? If I am understanding your example and the docs correctly, is this saying sample sched_switch stacks every ~50Hz? If so, this is ok - but we would actually want to be able to configure a callstack on for every certain type of sched event, and leave the frequency sampling for CPU Sampling.

From  https://perfetto.dev/docs/reference/trace-config-proto#PerfEvents.Timebase
PerfEvents.Timebase - frequency - Per-cpu sampling frequency in Hz, as requested from the kernel.

**Comment by primiano** [comment]
> When you say extremely new feature, w.r.t trace capture does that imply a certain version of Android or Linux?

The feature itself sits on top of [`man perf_event_open`](https://man7.org/linux/man-pages/man2/perf_event_open.2.html) which is a quite well established Linux kernel API (the same that the `perf` tool and many other use).

On Android, this depends on a daemon (traced_perf) which we introduced only in Android 12.
On Linux, assuming root or elevated capabilities, it works with a recent version of perfetto (see CHANGELOG)

There are a bunch of rough edges. For instance for now the symbolizer can do only dwarf-based unwinding, so requires some debugging symbols to be in the binary.

 > How does the frequency fit into this? 

Yeah I know where the doubt comes from.
There are actually two field you can use here, `period` and `frequency`. For my example above, "period" would have been better (Which is really the classic overflow count).

The comments [here](https://github.com/google/perfetto/blob/f4ffdc1c0d10c444b6ca626319d6cdb20f8edae4/protos/perfetto/common/perf_events.proto#L28) say it all.

TL;DR
- In cases like this you probably want to use period (i.e. sample every N events).
- If you use frequency, the kernel will do some adaptive sampling trying to estimate the frequency of events based on the past rate and give you an outbound rate that somewhat matches that frequency. But that code path seem to have various rough edges.

**Comment by ivberg** [comment]
BTW, Ideally we would want to be able to have a config that says capture EVERY sched_switch event AND also CPU sample every ~1ms or some frequency.

**Comment by ivberg** [comment]
> Yep I think that-s the issue.
Yep - that's it! Sorry if it was confusing. At least it's possible to gather sched_switch stacks at all, so we are halfway there ;) We would eventually require these stacks for meaningful critical path analysis. If the critical path (of a thread executing a scenario) includes CPU usage, which it often does - then the CPU Sampling stacks are also usually required to also provide useful data and root cause to a perf analyst.

**Comment by benteitler** [comment]
Was there ever a resolution to recording both scheduling events and time based sampling at the same time?

**Comment by LalitMaganti** [comment]
Just wanted to add a comment here: for local usage, you should now be able to use simpleperf + the recently added Gecko import support to import perf profiles with this feature. We sill don't support this through Perfetto natively but maybe it will unblock some usecases.

**Comment by LalitMaganti** [comment]
With https://perfetto.dev/docs/case-studies/scheduling-blockages we now demonstrate this is possible with with Perfetto as well (though with how unwinding works, it's totally impossible we'll be able to capture callstacks on *every* sched_swtich). That post has lots of useful info about how this works, how you can tune it etc.

**Comment by mstange** [comment]
Neat. Is there a follow-up issue about adding the ability to choose in-kernel frame pointer unwinding of the user stack?

**Comment by LalitMaganti** [comment]
Frame pointer unwinding landed in https://github.com/google/perfetto/commit/ab3bf51c8300eb01b00c5d0029a87f1d34dfba40 so it's been supported for quite some time.

**Comment by mstange** [comment]
That's not what I'm after; I've filed #3300 with more details.
